Few-shot learning in practice: GPT-NEO and the Accelerated Inference API

philschmid opened the pull request:
This blog post is about how to use `GPT-NEO` with the Accelerated Inference API. It tries to be understandable for now NLP Experts. 
> _keep in mind in the messaging that the inference-api targets ML engineer/infrastructure/software engineers who probably don't know what zero-shot is / might not be using colab too much I suspect. - clem_

The main focus point is the `few-shot-inference-widget`, which is a small playground for `GPT-NEO` and the inference-API.

@julien-c how can I validate if the widget is working? I have the hub locally set up.

julien-c commented:
> @julien-c how can I validate if the widget is working? I have the hub locally set up.

does it work locally?

Narsil commented:
https://github.com/huggingface/api-inference/pull/410 (FYI, it's quite important I've seen quite a bit of frustration from users trying to get to use Neo after it somehow crashed, this should help)

yjernite commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+
+## What is GPT-Neo?
+
+GPT⁠-⁠Neo is a family of transformer-based language models from [EleutherAI](https://www.eleuther.ai/projects/gpt-neo/) based on GPT architecture. [EleutherAI](https://www.eleuther.ai) primary goal is to train an equivalent model to the full-sized GPT⁠-⁠3 and make it available to the public under an open license.
```
```suggestion
GPT⁠-⁠Neo is a family of transformer-based language models from [EleutherAI](https://www.eleuther.ai/projects/gpt-neo/) based on the GPT architecture. [EleutherAI](https://www.eleuther.ai)'s primary goal is to train a model that is equivalent in size to GPT⁠-⁠3 and make it available to the public under an open license.
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+---
+title: 'Few-shot learning in practice: GPT-NEO and the Accelerated Inference API'
```
```suggestion
title: 'Few-shot learning in practice: GPT-Neo and the 🤗 Accelerated Inference API'
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+---
+
+<h1>
+    Few-shot learning in practice: GPT-NEO and the Accelerated Inference API
```
```suggestion
    Few-shot learning in practice: GPT-Neo and the 🤗 Accelerated Inference API
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+
+Few-Shot learning refers to the practice of feeding a learning model (language model) with a very small amount of data, contrary to the normal practice of fine-tuning using a large amount of data.
+
+This technique was in the past mostly used in computer vision, but with the latest and greatest Language models, like [OpenAIs GPT-3](https://openai.com/blog/gpt-3-apps/) and [EleutherAI GPT-Neo](https://www.eleuther.ai/projects/gpt-neo/), it is now also used in Natural Language Processing (NLP). 
```
```suggestion
This technique has been mostly used in computer vision, but with some of the latest Language Models, like [EleutherAI GPT-Neo](https://www.eleuther.ai/projects/gpt-neo/) and [OpenAI GPT-3](https://openai.com/blog/gpt-3-apps/), we can now use it in Natural Language Processing (NLP). 
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+
+## What is few-shot learning?
+
+Few-Shot learning refers to the practice of feeding a learning model (language model) with a very small amount of data, contrary to the normal practice of fine-tuning using a large amount of data.
```
```suggestion
Few-Shot Learning refers to the practice of feeding a machine learning model with a very small amount of training data to guide its predictions, like a few examples at inference time, as opposed to standard fine-tuning techniques which require a relatively large amount of training data for the pre-trained model to adapt to the desired task with accuracy.
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+
+This technique was in the past mostly used in computer vision, but with the latest and greatest Language models, like [OpenAIs GPT-3](https://openai.com/blog/gpt-3-apps/) and [EleutherAI GPT-Neo](https://www.eleuther.ai/projects/gpt-neo/), it is now also used in Natural Language Processing (NLP). 
+
+In NLP few-shot learning is used together with a large Language Model. These Language Models learned to perform a large number of tasks implicitly in order to reduce perplexity during their pre-training on large text datasets. This enables the Language model to generalize/understand unseen (but related) tasks with just a few examples.
```
```suggestion
In NLP, Few-Shot Learning can be used with Large Language Models, which have learned to perform a wide number of tasks implicitly during their pre-training on large text datasets. This enables the model to generalize, that is to understand related but previously unseen tasks, with just a few examples.
```

jeffboudier replied:
Here I wonder if it would make sense to plug Teven/Sasha's paper, e.g.
If you're wondering how many training examples is a prompt worth, well, so did we, and Hugging Face researchers Teven Le Scao and Sasha Rush ran the numbers in [this paper](https://arxiv.org/abs/2103.08493).

philschmid replied:
Great idea! added it. @TevenLeScao & @srush do you have any "profile" links that I can add to your names?

julien-c replied:
just use their hf.co profiles

TevenLeScao replied:
Great idea, thanks a bunch! Maybe link to Twitter (https://twitter.com/Fluke_Ellington, https://twitter.com/srush_nlp) since the hf.co profiles are a bit sparse right now

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+
+In NLP few-shot learning is used together with a large Language Model. These Language Models learned to perform a large number of tasks implicitly in order to reduce perplexity during their pre-training on large text datasets. This enables the Language model to generalize/understand unseen (but related) tasks with just a few examples.
+
+Few-Shot NLP examples consist of three main components: 
+
+- **Task Description:** A short description of what the model should do, e.g. translate English to France
+- **Examples:** A few shot examples demonstrating/showing the model what it should do, e.g. sea otter => loutre de mer
+- **Prompt:** The beginning of example where the model should generate the missing text for completion, e.g. cheese =>
```
```suggestion
Few-Shot NLP examples consist of three main components: 

- **Task Description**: A short description of what the model should do, e.g. "Translate English to French"
- **Examples**: A few examples showing the model what it is expected to predict, e.g. "sea otter => loutre de mer"
- **Prompt**: The beginning of a new example, which the model should complete by generating the missing text, e.g. "cheese => "
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+![few-shot-prompt](assets/22_few_shot_learning_gpt_neo_and_inference_api/few-shot-prompt.png)  
+<small>Image from <a href="https://arxiv.org/abs/2005.14165" target="_blank">Language Models are Few-Shot Learners</a></small>
+
+Creating these few-shot examples can be tough since you need to elaborate the “task” you want to perform to the model. A common problem is that models, especially smaller ones seem to be very sensitive to the way the example is written.
```
```suggestion
Creating these few-shot examples can be tricky, since you need to articulate the “task” you want the model to perform through them. A common issue is that models, especially smaller ones, are very sensitive to the way the examples are written.
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+
+Creating these few-shot examples can be tough since you need to elaborate the “task” you want to perform to the model. A common problem is that models, especially smaller ones seem to be very sensitive to the way the example is written.
+
+OpenAI contributed in their [GPT-3 Paper](https://arxiv.org/abs/2005.14165) an observation that the few-shot prompting ability grows as the number of language model parameters grows.
```
```suggestion
OpenAI showed in the [GPT-3 Paper](https://arxiv.org/abs/2005.14165) that the few-shot prompting ability improves with the number of language model parameters.
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+
+## 🤗 Accelerated Inference API
+
+The inference API is our fully managed and hosted model serving product. The Inference API is capable of serving all of the 10,000+ public models from our hub or your own private models without any adjustments. The API includes acceleration on CPU and GPU with performance up 100x compared to basic transformer implementations.
```
```suggestion
The Accelerated Inference API is our hosted service to run inference on any of the 10,000+ models publicly available on the 🤗 Model Hub, or your own private models, via simple API calls. The API includes acceleration on CPU and GPU with [up to 100x speedup](https://huggingface.co/blog/accelerated-inference) compared to out of the box deployment of Transformers.
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+
+The inference API is our fully managed and hosted model serving product. The Inference API is capable of serving all of the 10,000+ public models from our hub or your own private models without any adjustments. The API includes acceleration on CPU and GPU with performance up 100x compared to basic transformer implementations.
+
+If you want to use `GPT-Neo` for a few-shot prediction programmatically in your app you can use the following snippet. You can your API Token [here](https://huggingface.co/settings/token) and If you don't have an account you can get started [here](https://huggingface.co/pricing).
```
```suggestion
To integrate Few-Shot Learning predictions with `GPT-Neo` in your own apps, you can use the 🤗 Accelerated Inference API with the code snippet below. You can find your API Token [here](https://huggingface.co/settings/token), if you don't have an account you can get started [here](https://huggingface.co/pricing).
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,119 @@
+
+---
+
+## Unfair advantage
+
+Our partnerships and open-source collaborations with hardware and cloud vendors like Intel, NVIDIA, Qualcomm, Amazon, and Microsoft enable us to tune our API infrastructure with the latest hardware optimization techniques.
+
+As Machine Learning Engineers at Hugging Face we certainly have an unfair advantage sitting in the same (virtual) offices as the 🤗 Transformers and 🤗 Tokenizers maintainers.
+
+### When You/your company should use the Accelerated Inference API:
+
+- When your idea/product wants to benefit from rapid development through the easy integration of all ~11 000 models of the [Hugging Face Hub](https://huggingface.co/models).
+- When you want to get up to [100x performance optimizations](https://huggingface.co/blog/accelerated-inference) out of the box.
+- When you don't have in-house expertise on how to host billion parameter large NLP models
+- When you want transparency which model is used for inference compared to, e.g. [Amazon Comprehend](https://aws.amazon.com/de/comprehend/) or [Google Cloud NLP](https://cloud.google.com/natural-language)
+
+---
+
+If you want to feel the speed on our infrastructure, start a [free trial](https://huggingface.co/pricing) and we’ll get in touch. If you want to benefit from our experience optimizing inference on your own infrastructure participate in our [🤗 Expert Acceleration Program](https://huggingface.co/support).
```
I would remove this section, it feels more marketing than blog post.
What could be interesting to add IMO would be practical tips on how to design prompts that work well, and mistakes to avoid, which could simply be a commentary of the default examples you have set up in the widget

jeffboudier reviewed (commented):
I left some copy tweaks as suggestions. I think in the last part it would be interesting to discuss the prompting examples you set up in the widget, give some tips on what makes a good prompt, common mistakes.

philschmid commented:
@jeffboudier i removed the marketing section and added some practical insights with visual, see attached screenshot. 
> I would remove this section, it feels more marketing than blog post.
> What could be interesting to add IMO would be practical tips on how to design prompts that work well, and mistakes to avoid, which could simply be a commentary of the default examples you have set up in the widget. This could allow a natural plug to the EAP.

![image](https://user-images.githubusercontent.com/32632186/119480661-2fb08e80-bd52-11eb-88f6-2b18c0901101.png)

jeffboudier commented:
I love the examples @philschmid !
Unlike the other concepts, `Temperature` is not self-explanatory, I think could be quickly defined.
And I wonder if you had the chance to sync with @lewtun on this as I saw on Slack he just worked on the few-shot learning chapter of the book.

lewtun commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,123 @@
+
+<script defer src="https://gpt-neo-accelerated-inference-api.s3-eu-west-1.amazonaws.com/fewShotInference.js"></script>
+<few-shot-inference-widget ></few-shot-inference-widget>
+
```
I wonder if we need an introductory paragraph to tell the reader what we'll be covering in this blog post? For example, something like:

In many NLP applications, the amount of available labeled data is often a barrier to producing a high-performing model. However, thanks to recent developments in scaling-up language models, it now becoming possible to overcome this limitation by simply providing a handful of examples at inference time, a technique known as _Few-Shot Learning_. In this blog post, we'll explain what Few-Shot Learning is and explore how a large language model called GPT-Neo and the 🤗 Accelerated Inference API can be used to generate your own predictions.

yjernite commented:
This is looking good :) 

What's the timeline for release? I think it's really important to add a section on limitations which includes:
- A warning not to use the system in fully automated systems that make decisions on behalf of users
- Some examples of not-so-great generations (because if we don't put them forward, someone else will)
- If we can frame the above as "this is bad but not as bad as such and such GPT-3 example", that's even better

Here are examples of model cards from Microsoft with good limitations sections
- [Sentiment detection](https://docs.microsoft.com/en-us/legal/cognitive-services/text-analytics/transparency-note-sentiment-analysis#considerations-when-choosing-a-use-case)
- [Named Entity Recognition](https://docs.microsoft.com/en-us/legal/cognitive-services/text-analytics/transparency-note-named-entity-recognition#considerations-when-choosing-a-use-case)

I can spend a couple of hours on this section depending on the prefer but I think it would be better if a few of us worked on it together (cc @jeffboudier and anyone who wants to help with this aspect :) )

philschmid commented:
I would merge/release the blog post tomorrow. I checked if the `inference-widget` works locally, it does. 
![Bildschirmfoto 2021-06-01 um 17 09 41](https://user-images.githubusercontent.com/32632186/120347599-b9d19780-c2fc-11eb-866b-a1ea358657b5.png)

yjernite commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,125 @@
+
+In the example, you can see how important it is to define your hyperparameter. These can make the difference between solving your task or failing miserably.
+
+
```
```suggestion
---
## Considerations for Using the System

Just as few-shot learning can enable new uses of machine learning and natural language processing in cases where the lack of annotated data was previously prohibitive, this new paradigm also presents unique pitfalls that need to be taken into account when designing uses cases.

The default `Sentiment Analysis` setting provided in the widget can help us outline some of these risks. After seeing three examples of sentiment classification, the model makes the following predictions. With a temperature parameter set as 0.1, four times out of five, we get:
> ###
> Tweet: "I'm a tall blond person"
> Sentiment: Positive
Whereas we get (also four times out of five with the same parameters):
> ###
> Tweet: "I'm a disabled happy person"
> Sentiment: Negative

Imagine for example that you are using this sentiment analysis model to aggregate reviews of products on an online shopping website: what can go wrong in this case? One possible outcome would be that items that are useful to people with disabilities would be systematically down-ranked and harder to find regardless of their actual quality, which is a form of automated discrimination.  We refer readers to the ACL 2020 paper [Social Biases in NLP Models as Barriers for Persons with Disabilities](https://www.aclweb.org/anthology/2020.acl-main.487.pdf) for more information. In particular, the few-shot learning paradigm relies more directly on information and associations the model has picked up from its pre-training phase, making it even more sensitive to these categories of failures.

Now that we have gone over some of the risks, let us consider some good practices that may help minimize them when choosing a use case.

### Considerations for Using the System: Do
When designing a use case that relies on this system, **do**:

- Make sure that people know which parts of their user experience depend on the outputs of the system
- If possible, give users the ability to opt-out if they so choose
- Provide a mechanism for users to give feedback on the model decision and overwrite it if they are unhappy with it
- Monitor the feedback, especially model failures, for groups of users that may be disproportionately affected

### Considerations for Using the System: Don't
When designing a use case that relies on this system, **do not**:

- Use the model to automatically make decisions for or about a user without an opportunity for human input or redress

In particular, note that the right of users to at least be provided an explanation for automatic decisions made about them is enshrined in several regulations, such as the European [GDPR](https://gdpr-info.eu/).

```

philschmid replied:
@yjernite I would rather not add this, because as @clmnt already said it is not aimed at the target group and scares engineers/developers rather than helps them. 
Thoughts on that @jeffboudier ?

yjernite replied:
Happy to work on the phrasing and edit the paragraph, but I do think it should be there if only to tell people what they'd be paying for with the EAP :)

jeffboudier replied:
I worked on the section to tighten it up, without removing any of the content. I didn't mean to replace it 😅, the original text is below:

---
## Considerations for Using the System
Just as few-shot learning can enable new uses of machine learning and natural language processing in cases where the lack of annotated data was previously prohibitive, this new paradigm also presents unique pitfalls that need to be taken into account when designing uses cases.
The default `Sentiment Analysis` setting provided in the widget can help us outline some of these risks. After seeing three examples of sentiment classification, the model makes the following predictions. With a temperature parameter set as 0.1, four times out of five, we get:
> ###
> Tweet: "I'm a tall blond person"
> Sentiment: Positive
Whereas we get (also four times out of five with the same parameters):
> ###
> Tweet: "I'm a disabled happy person"
> Sentiment: Negative
Imagine for example that you are using this sentiment analysis model to aggregate reviews of products on an online shopping website: what can go wrong in this case? One possible outcome would be that items that are useful to people with disabilities would be systematically down-ranked and harder to find regardless of their actual quality, which is a form of automated discrimination.  We refer readers to the ACL 2020 paper [Social Biases in NLP Models as Barriers for Persons with Disabilities](https://www.aclweb.org/anthology/2020.acl-main.487.pdf) for more information. In particular, the few-shot learning paradigm relies more directly on information and associations the model has picked up from its pre-training phase, making it more sensitive to these categories of failures.
Now that we have gone over some of the risks, let us consider some good practices that may help minimize them when choosing a use case.
### Considerations for Using the System: Do
When designing a use case that relies on this system, **do**:
- Make sure that people know which parts of their user experience depend on the outputs of the system
- If possible, give users the ability to opt-out if they so choose
- Provide a mechanism for users to give feedback on the model decision and overwrite it if they are unhappy with it
- Monitor the feedback, especially model failures, for groups of users that may be disproportionately affected
### Considerations for Using the System: Don't
When designing a use case that relies on this system, **do not**:
- Use the model to automatically make decisions for or about a user without an opportunity for human input or redress
In particular, note that the right of users to at least be provided an explanation for automatic decisions made about them is enshrined in several regulations, such as the European [GDPR](https://gdpr-info.eu/).

yjernite replied:
Sounds good! I like the edits + the reference to EAP at the end. @philschmid does that look better to you?

yjernite reviewed (changes requested):
I added a paragraph on `Considerations for Using the System`

@clmnt @jeffboudier it still needs something at the end. How do we feel about adding something along the lines of "We can help you navigate these concerns within our EAP"?

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,122 @@
+    </a>
+</div>
+
+In many Machine Learning applications, the amount of available labeled data is a barrier to producing a high-performing model. The latest developments in NLP shown that you can overcome this limitation by providing a few examples at inference time with a large language model - a technique known as Few-Shot Learning. In this blog post, we'll explain what Few-Shot Learning is and explore how a larg ...
```
```suggestion
In many Machine Learning applications, the amount of available labeled data is a barrier to producing a high-performing model. The latest developments in NLP show that you can overcome this limitation by providing a few examples at inference time with a large language model - a technique known as Few-Shot Learning. In this blog post, we'll explain what Few-Shot Learning is, and explore how a large language model called GPT-Neo, and the 🤗 Accelerated Inference API, can be used to generate your own predictions.
```

jeffboudier commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,122 @@
+
+---
+
+If you want to feel the speed on our infrastructure, start a [free trial](https://huggingface.co/pricing) and we’ll get in touch. If you want to benefit from our experience optimizing inference on your own infrastructure participate in our [🤗 Expert Acceleration Program](https://huggingface.co/support).
```
```suggestion
To use GPT-Neo or any Hugging Face model in your own application, you can [start a free trial](https://huggingface.co/pricing) of the 🤗 Accelerated Inference API.
If you need help mitigating bias in models and AI systems, or leveraging Few-Shot Learning, the 🤗 Expert Acceleration Program can [offer your team direct premium support from the Hugging Face team](https://huggingface.co/support).
```

philschmid commented:
I added the suggestions for "Considerations for Using the System". Here is a final screenshot of the rendered page

![screencapture-localhost-5564-blog-few-shot-learning-gpt-neo-and-inference-api-2021-06-03-16_58_05](https://user-images.githubusercontent.com/32632186/120666537-f8965780-c48c-11eb-8e92-70fb5388ad8f.png)

yjernite reviewed (approved):
Works for me, but I think @jeffboudier's changes got lost in the process unfortunately! I think that might be some Git web UI weirdness where maybe his editing my suggested change didn't propagate when they were committed?

https://github.com/huggingface/blog/pull/111#discussion_r643232280

philschmid commented:
@yjernite they are included

jeffboudier commented:
Good catch @yjernite 
@philschmid I had edited the section to condense it, see the edited comment, suggestion starting with "Responsible Use"  - could we update it? https://github.com/huggingface/blog/pull/111#discussion_r643232280

julien-c commented on few-shot-learning-gpt-neo-and-inference-api.md:
```diff
@@ -0,0 +1,153 @@
+            <span class="fullname">Philipp Schmid</span>
+        </div>
+    </a>
+</div>
```
we usually also add the button to edit on GitHub

philschmid replied:
Ah shit, i knew I delete the wrong thing. I would add it again and then we could also redeploy with yacines/jeffs changes
